### PR TITLE
GlobalOpt: Use the correct address space when creating a "*.init" global.

### PR DESCRIPTION
**Pull request by resistor**

### CODE REVIEW

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp [hunk 1]
```diff
@@ -950,7 +950,8 @@ OptimizeGlobalAddressOfAllocation(GlobalVariable *GV, CallInst *CI,
     new GlobalVariable(Type::getInt1Ty(GV->getContext()), false,
                        GlobalValue::InternalLinkage,
                        ConstantInt::getFalse(GV->getContext()),
-                       GV->getName()+".init", GV->getThreadLocalMode());
+                       GV->getName()+".init", GV->getThreadLocalMode(),
+                       GV->getType()->getAddressSpace());
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/GlobalOpt.cpp:954]:
```suggestion
                       GV->getAddressSpace());
```

File: llvm/test/Transforms/GlobalOpt/malloc-promote-addrspace.ll [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+; CHECK:       3:
+; CHECK-NEXT:    [[TMP4:%.*]] = phi i32 [ -1, [[TMP1]] ], [ 1, [[TMP2]] ]
+; CHECK-NEXT:    ret i32 [[TMP4]]
+;
+  %1 = load ptr addrspace(200), ptr addrspace(200) @g, align 8
+  call void @f2();
+  %2 = icmp eq ptr addrspace(200) %1, null
+  br i1 %2, label %3, label %4
+
+3:                                          ; preds = %0
```
Comment (arsenm) [inline llvm/test/Transforms/GlobalOpt/malloc-promote-addrspace.ll:28]: Use named values in tests

Comment (resistor) [reply llvm/test/Transforms/GlobalOpt/malloc-promote-addrspace.ll]: Done

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp (hunk 1 above)
Comment (resistor) [reply llvm/lib/Transforms/IPO/GlobalOpt.cpp]: Done

**Review by arsenm — APPROVED** [review]

**Comment by resistor** [comment]
Tagging @jrtc27 and @arichardson for awareness on CHERI-derived PRs

**Pull request merged**